Camera image provider & carousel widget

sandreae opened the pull request:
A few image capture related widgets/screens introduced here which are all related to creating a new sighting. 

1. Image provider inherited widget which exposing methods for: 
    - capturing images from camera and filesystem
    - accessing all currently captured images
    - removing an image by index
    - removing all captured images
2. Carousel widget for displaying and deleting captured images
3. Expandable floating action button for picking images from camera or gallery

NB: I tried to get things as close visually to our designs as possible, but I don't consider it finished. For example, we're missing buttons on the carousel, and the image positioning is a bit lame. But worth getting it merged in this state for now I think.  

https://github.com/p2panda/meli/assets/24291662/813fa102-bd2f-4d3d-b3bd-5b6399853952

adzialocha commented on packages/app/lib/ui/screens/create_sighting.dart:
```diff
@@ -21,188 +18,97 @@ class CreateNewScreen extends StatefulWidget {
+          _initialImagePickAttemptComplete = true;
+
+          if (snapshot.hasError) {
+            print('ImageProvider Error: ${snapshot.error}');
```
This is when the camera fails, right? Is it then still possible to add images via the gallery? If even this doesn't work we might just want to redirect the user to the main screen?

adzialocha commented on packages/app/lib/ui/widgets/image_provider.dart:
```diff
@@ -0,0 +1,131 @@
+  @override
+  bool updateShouldNotify(MeliCameraProviderInherited oldWidget) {
+    bool result = (images != oldWidget.images);
+    print(result);
```
Remove print

adzialocha commented on packages/app/lib/ui/widgets/sighting_form.dart:
```diff
@@ -1,149 +1,143 @@
+  }
+}
+
+class SightingImagesCarousel extends StatelessWidget {
```
This could be in its own file

adzialocha commented on packages/app/lib/ui/widgets/sighting_form.dart:
```diff
@@ -1,149 +1,143 @@
+class SightingImagesCarousel extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    final cameraImageProvider = MeliCameraProviderInherited.of(context);
+    final showDelete = !cameraImageProvider.isEmpty();
+    final images = !cameraImageProvider.isEmpty()
+        ? cameraImageProvider.getAll().map((file) {
+            return Image.file(file);
+          })
+        : [Image.asset(PLACEHOLDER_IMG)];
```
Does it need to be so closely tied to the camera functionality? I would assume that the carousel is just a "simple" widget displaying any image it gets from the outside, not needing to know where they came from or how they are managed.

adzialocha commented on packages/app/lib/ui/widgets/sighting_form.dart:
```diff
@@ -1,149 +1,143 @@
+                    shape: MaterialStateProperty.all<CircleBorder>(
+                        CircleBorder())),
+                onPressed: () {
+                  cameraImageProvider.removeAt(index);
```
Also here it should rather provider a generic callback that an image was deleted. The camera provider is too specific here (we will need this carousel also in other views, independent of the camera).

adzialocha commented on packages/app/pubspec.yaml:
```diff
@@ -31,6 +32,7 @@ dev_dependencies:
 flutter:
   assets:
     - assets/
+    - assets/images/
```
Nice bee 💘

adzialocha commented on packages/app/lib/app.dart:
```diff
@@ -16,7 +17,7 @@ class MeliApp extends StatelessWidget {
     return GraphQLProvider(
         client: client,
-        child: MaterialApp.router(
+        child: MeliCameraProvider(MaterialApp.router(
```
I've just stumpled upon `MultiProvider`, might be nicer to read:

```
MultiProvider(
  providers: [
    ChangeNotifierProvider(builder: (_) => FirstProvider()),
    ChangeNotifierProvider(builder: (_) => SecondProvider()),
  ],
  child: someWidget,
)
```

sandreae replied:
I'll wait and see if we wanna move where the wrapping happens before changing this. I like the pattern in any case :+1:

sandreae replied:
Unless I'm mistaken we'll need to adapt our current providers (which extend inherited widget) in order to use this. It is nice though, so happy for this refactor to happen in a follow-up step.

adzialocha commented on packages/app/lib/ui/widgets/image_provider.dart:
```diff
@@ -0,0 +1,131 @@
+
+    XFile? imageFile = await _imagePicker.pickImage(
+        source: ImageSource.camera,
+        imageQuality: 50,
```
Do we want to increase the image quality? If it is used for scientific purposes? How about 80?

sandreae replied:
Yeh, will bump to 80!

adzialocha commented on packages/app/lib/ui/widgets/image_provider.dart:
```diff
@@ -0,0 +1,131 @@
+  // Pick photos from the gallery.
+  Future<void> pickFromGallery() async {
+    List<XFile> imageFiles =
+        await _imagePicker.pickMultiImage(imageQuality: 50);
```
Same here (image quality)

adzialocha reviewed (changes requested):
So cool! Think it's almost ready to go, also the provider pattern looks good. It would be probably also fine to just have it for that particular view, but I don't mind it there.

I have some more minor questions and requests (see comments above), mostly around the carousel.

adzialocha commented on packages/app/lib/app.dart:
```diff
@@ -1,5 +1,6 @@
 // SPDX-License-Identifier: AGPL-3.0-or-later
 
+import 'package:app/ui/widgets/image_provider.dart';
```
.. this also goes for a few other places in this PR

adzialocha reviewed (commented):
Ignore this comment one (I can't delete it) 😆

sandreae commented on packages/app/lib/ui/screens/create_sighting.dart:
```diff
@@ -21,188 +18,97 @@ class CreateNewScreen extends StatefulWidget {
+                  await createSighting(datetime, 0.0, 0.0, [], null, null,
+                      "Some comment about this sighting");
+
+                  // Remove captured images from the provider
+                  cameraImageProvider.removeAll();
```
If we wanna do this in a `dispose` method I think the place for it would be in the `MeliCameraProviderState` widget. As I understand it you normally want to access inherited widgets in the `build` method.

Only problem there is that as the whole app is wrapped in the image provider then dispose won't get called when navigating away from the create sightings screen. 

Think this would require moving the wrapping down a little further in the tree.

adzialocha replied:
Why not in the dispose method of the `CreateSighting` widget?

adzialocha replied:
[device-2023-10-02-173002.webm](https://github.com/p2panda/meli/assets/1822473/2c10a4a1-ddcf-47fb-9c4a-f4a74722f74b)

I've made some experiments, It is possible to enter the view and see old pictures, either when cancelling the camera view or pressing the back button in the Android bar. I think if we somehow can clear the images in the provider whenever the create sighting view is exited, that would be ideal.

adzialocha replied:
Think the behaviour we want is: You can only enter the quick sightings view when you've selected _at least_ one image. When you cancel or press back the user probably just wants to stop going through the flow.

sandreae replied:
Nice, made this change.

With this there is no way to get to the create sighting form without taking a picture (it was janky before, but there was a way), not sure if that's the behaviour we want. Let's see as we use the app a bit, might be that we have to have an expandable fab on the homepage too.

adzialocha commented on packages/app/lib/ui/widgets/image_carousel.dart:
```diff
@@ -0,0 +1,81 @@
+  Widget build(BuildContext context) {
+    return Container(
+      padding: EdgeInsets.symmetric(vertical: 5),
+      child: CarouselSlider(
```
I would somehow assume that the items "within" the box would move, not the whole box itself when I slide from one to the next image. Would it be possible to have the border around the carousel instead?

adzialocha reviewed (changes requested):
We're getting closer! :-) Some more things came up but I think that's almost it!

1) I've observed this problem here:

[device-2023-10-04-093012.webm](https://github.com/p2panda/meli/assets/1822473/8cbee324-612f-435c-88ec-7b0fd0a5df4f)

2) In the Figma Design we have arrows (previous / next) for the carousel. I think it doesn't explain by itself that it is a carousel so we either have these dots for each image indicating that there's more and / or the next / previous buttons? I'd prefer the buttons as we don't have much space at the bottom for the dots (the delete button is there).

3) Should we give the delete functionality a confirmation dialogue? It goes quite quick.

4. Some sort of confirmation (snack bar) would also be nice in case a delete animation is too much work? It would make clearer that I've just deleted something.

adzialocha commented:
> In the Figma Design we have arrows (previous / next) for the carousel. I think it doesn't explain by itself that it is a carousel so we either have these dots for each image indicating that there's more and / or the next / previous buttons? I'd prefer the buttons as we don't have much space at the bottom for the dots (the delete button is there).

Ah, forgot that you've mentioned this in your PR description. Up to you!

sandreae commented:
> I've observed this problem here (deleting all images brings me to the loading state):
oops, fixed

> In the Figma Design we have arrows (previous / next) for the carousel. I think it doesn't explain by itself that it is a carousel so we either have these dots for each image indicating that there's more and / or the next / previous buttons? I'd prefer the buttons as we don't have much space at the bottom for the dots (the delete button is there).

Made it auto-play so it's clear this is a carousel, I'd prefer to perfect the style in following steps.

> Should we give the delete functionality a confirmation dialogue before deletion?

yep, done

> Some sort of confirmation (snack bar) after deletion would also be nice in case a delete animation is too much work? It would make clearer that I've just deleted something.

yep, done
